Index: kdelibs-4.10.0/solid/solid/backends/udisks2/udisksopticaldisc.cpp
===================================================================
--- kdelibs-4.10.0.orig/solid/solid/backends/udisks2/udisksopticaldisc.cpp
+++ kdelibs-4.10.0/solid/solid/backends/udisks2/udisksopticaldisc.cpp
@@ -25,8 +25,6 @@
 #include <fcntl.h>
 
 #include <QtCore/QFile>
-#include <QtCore/QMap>
-#include <QtCore/QMutexLocker>
 #include <QtDBus/QDBusConnection>
 
 #include "../shared/udevqt.h"
@@ -35,10 +33,6 @@
 #include "udisksopticaldisc.h"
 #include "soliddefs_p.h"
 
-typedef QMap<QByteArray, Solid::OpticalDisc::ContentTypes> ContentTypesCache;
-SOLID_GLOBAL_STATIC(ContentTypesCache, cache)
-SOLID_GLOBAL_STATIC(QMutex, cacheLock)
-
 // inspired by http://cgit.freedesktop.org/hal/tree/hald/linux/probing/probe-volume.c
 static Solid::OpticalDisc::ContentType advancedDiscDetect(const QByteArray & device_file)
 {
@@ -173,7 +167,7 @@ out:
 using namespace Solid::Backends::UDisks2;
 
 OpticalDisc::OpticalDisc(Device *dev)
-    : StorageVolume(dev), m_needsReprobe(true), m_cachedContent(Solid::OpticalDisc::NoContent)
+    : StorageVolume(dev)
 {
     UdevQt::Client client(this);
     m_udevDevice = client.deviceByDeviceFile(device());
@@ -181,8 +175,6 @@ OpticalDisc::OpticalDisc(Device *dev)
     /*qDebug() << "\tProperties:" << */ m_udevDevice.deviceProperties(); // initialize the properties DB so that it doesn't crash further down, #298416
 
     m_drive = new Device(m_device->prop("Drive").value<QDBusObjectPath>().path());
-    QDBusConnection::systemBus().connect(UD2_DBUS_SERVICE, m_drive->udi(), DBUS_INTERFACE_PROPS, "PropertiesChanged", this,
-                                         SLOT(slotDrivePropertiesChanged(QString,QVariantMap,QStringList)));
 }
 
 OpticalDisc::~OpticalDisc()
@@ -246,22 +238,12 @@ Solid::OpticalDisc::DiscType OpticalDisc
 Solid::OpticalDisc::ContentTypes OpticalDisc::availableContent() const
 {
     if (isBlank()) {
-        m_needsReprobe = false;
         return Solid::OpticalDisc::NoContent;
     }
 
-    if (m_needsReprobe) {
-        QMutexLocker lock(cacheLock);
-
         const QByteArray deviceFile = m_device->prop("Device").toByteArray();
 
-        if (cache->contains(deviceFile)) {
-            m_cachedContent = cache->value(deviceFile);
-            m_needsReprobe = false;
-            return m_cachedContent;
-        }
-
-        m_cachedContent = Solid::OpticalDisc::NoContent;
+    Solid::OpticalDisc::ContentTypes m_cachedContent = Solid::OpticalDisc::NoContent;
         const bool hasData = m_drive->prop("OpticalNumDataTracks").toUInt() > 0;
         const bool hasAudio = m_drive->prop("OpticalNumAudioTracks").toUInt() > 0;
 
@@ -272,25 +254,9 @@ Solid::OpticalDisc::ContentTypes Optical
         if ( hasAudio )
             m_cachedContent |= Solid::OpticalDisc::Audio;
 
-        m_needsReprobe = false;
-        cache->insert(deviceFile, m_cachedContent);
-    }
-
     return m_cachedContent;
 }
 
-void OpticalDisc::slotDrivePropertiesChanged(const QString &ifaceName, const QVariantMap &changedProps, const QStringList &invalidatedProps)
-{
-    Q_UNUSED(ifaceName);
-
-    if (changedProps.keys().contains("Media") || invalidatedProps.contains("Media")) {
-        QMutexLocker lock(cacheLock);
-        m_needsReprobe = true;
-        m_cachedContent = Solid::OpticalDisc::NoContent;
-        cache->remove(m_device->prop("Device").toByteArray());
-    }
-}
-
 QString OpticalDisc::media() const
 {
     return m_drive->prop("Media").toString();
Index: kdelibs-4.10.0/solid/solid/backends/udisks2/udisksopticaldisc.h
===================================================================
--- kdelibs-4.10.0.orig/solid/solid/backends/udisks2/udisksopticaldisc.h
+++ kdelibs-4.10.0/solid/solid/backends/udisks2/udisksopticaldisc.h
@@ -52,13 +52,8 @@ public:
     virtual Solid::OpticalDisc::DiscType discType() const;
     virtual Solid::OpticalDisc::ContentTypes availableContent() const;
 
-private Q_SLOTS:
-    void slotDrivePropertiesChanged(const QString & ifaceName, const QVariantMap & changedProps, const QStringList & invalidatedProps);
-
 private:
     QString media() const;
-    mutable bool m_needsReprobe;
-    mutable Solid::OpticalDisc::ContentTypes m_cachedContent;
     Device * m_drive;
     UdevQt::Device m_udevDevice;
 };
