Index: kdelibs-4.10.0/solid/solid/backends/udisks2/udisksopticaldisc.cpp
===================================================================
--- kdelibs-4.10.0.orig/solid/solid/backends/udisks2/udisksopticaldisc.cpp
+++ kdelibs-4.10.0/solid/solid/backends/udisks2/udisksopticaldisc.cpp
@@ -176,6 +176,8 @@ OpticalDisc::OpticalDisc(Device *dev)
     m_drive = new Device(m_device->prop("Drive").value<QDBusObjectPath>().path());
     QDBusConnection::systemBus().connect(UD2_DBUS_SERVICE, m_drive->udi(), DBUS_INTERFACE_PROPS, "PropertiesChanged", this,
                                          SLOT(slotDrivePropertiesChanged(QString,QVariantMap,QStringList)));
+
+    m_prober = new DiskScannerInterface(this);
 }
 
 OpticalDisc::~OpticalDisc()
@@ -250,7 +252,7 @@ Solid::OpticalDisc::ContentTypes Optical
 
         if ( hasData ) {
             m_cachedContent |= Solid::OpticalDisc::Data;
-            m_cachedContent |= advancedDiscDetect(m_device->prop("Device").toByteArray());
+            m_cachedContent |= m_prober->getContent(m_device->udi(), m_device->prop("Device").toString());
         }
         if ( hasAudio )
             m_cachedContent |= Solid::OpticalDisc::Audio;
Index: kdelibs-4.10.0/solid/solid/backends/udisks2/udisksopticaldisc.h
===================================================================
--- kdelibs-4.10.0.orig/solid/solid/backends/udisks2/udisksopticaldisc.h
+++ kdelibs-4.10.0/solid/solid/backends/udisks2/udisksopticaldisc.h
@@ -29,6 +29,8 @@
 #include "udisksstoragevolume.h"
 #include "udisksdevice.h"
 
+#include "diskscannerinterface.h"
+
 namespace Solid
 {
 namespace Backends
@@ -61,6 +63,7 @@ private:
     mutable Solid::OpticalDisc::ContentTypes m_cachedContent;
     Device * m_drive;
     UdevQt::Device m_udevDevice;
+    mutable DiskScannerInterface *m_prober;
 };
 
 }
