--- solid/solid/backends/udisks2/udisksdevice.cpp.org	2013-02-07 09:12:48.071999541 +0100
+++ solid/solid/backends/udisks2/udisksdevice.cpp	2013-02-07 09:13:10.193999538 +0100
@@ -609,32 +609,6 @@
         {
             if ( drive.prop("Optical").toBool() )    // optical stuff
             {
-                bool isWritable = drive.prop("OpticalBlank").toBool();
-
-                const UDisks2::OpticalDisc disc(const_cast<Device*>(this));
-                Solid::OpticalDisc::ContentTypes availContent = disc.availableContent();
-
-                if (availContent & Solid::OpticalDisc::VideoDvd) // Video DVD
-                    return "media-optical-dvd-video";
-                else if ((availContent & Solid::OpticalDisc::VideoCd) || (availContent & Solid::OpticalDisc::SuperVideoCd)) // Video CD
-                    return "media-optical-video";
-                else if ((availContent & Solid::OpticalDisc::Data) && (availContent & Solid::OpticalDisc::Audio)) // Mixed CD
-                    return "media-optical-mixed-cd";
-                else if (availContent & Solid::OpticalDisc::Audio) // Audio CD
-                    return "media-optical-audio";
-                else if (availContent & Solid::OpticalDisc::Data) // Data CD
-                    return "media-optical-data";
-                else if ( isWritable )
-                    return "media-optical-recordable";
-                else
-                {
-                    if ( media.startsWith( "optical_dvd" ) || media.startsWith( "optical_hddvd" ) ) // DVD
-                        return "media-optical-dvd";
-                    else if ( media.startsWith( "optical_bd" ) ) // BluRay
-                        return "media-optical-blu-ray";
-                }
-
-                // fallback for every other optical disc
                 return "media-optical";
             }
 
