--- notmuch-0.18.1/bindings/ruby/extconf.rb.org	2014-10-07 08:54:14.728457776 +0200
+++ notmuch-0.18.1/bindings/ruby/extconf.rb	2014-10-07 08:55:07.539454213 +0200
@@ -5,27 +5,9 @@
 
 require 'mkmf'
 
-dir = File.join('..', '..', 'lib')
-
-# includes
-$INCFLAGS = "-I#{dir} #{$INCFLAGS}"
-
 # make sure there are no undefined symbols
-$LDFLAGS += ' -Wl,--no-undefined'
+$LDFLAGS += ' -Wl,--no-undefined -lnotmuch'
 
-def have_local_library(lib, path, func, headers = nil)
-  checking_for checking_message(func, lib) do
-    lib = File.join(path, lib)
-    if try_func(func, lib, headers)
-      $LOCAL_LIBS += lib
-    end
-  end
-end
-
-if not have_local_library('libnotmuch.so', dir, 'notmuch_database_create', 'notmuch.h')
-  exit 1
-end
-
 # Create Makefile
 dir_config('notmuch')
 create_makefile('notmuch')
